Taproot Compiler

SarcasticNastik opened the pull request:
This PR builds on top of #278. 

This is one in a series of PRs aimed to implement a feature-complete *Pay-to-Taproot* **P2Tr** compiler.
Specifically, this introduces a basic compilation for a given policy by collecting the leaf nodes of the *tree* generated by considering root-level disjunctive `OR`and using this to generate a `TapTree`. 

> Assuming that _duplicate keys_ are **NOT** possible even in different branches of the TapTree.

# Follow Up PRs
 
- #342  - Uses heuristic for tree-generation/ _merging_ algorithm while buillding `TapTree` to optimize over the *expected average total cost*. 
- #352 
- A future PR implementing enumerative strategies for optimizing `thresh` `k`-of-`n` and similar structures.

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -137,30 +143,109 @@ impl fmt::Display for PolicyError {
+
+        for key in concrete_keys.iter() {
+            if semantic_policy
+                .clone()
```
This algorithm is poor in memory performance because it calls clone for every single iteration. We can do a new tree traversal for every key we see. I think we will eventually need a better algorithm that avoids all these allocations. 

We can first focus on the correctness/code hygine of the PR, then address these efficiency issues.

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -137,30 +143,109 @@ impl fmt::Display for PolicyError {
-        )
-        .unwrap();
+        let tree =
+            Descriptor::new_tr(internal_key, Some(policy.compile_tr_policy().unwrap())).unwrap();
```
This should either propagate the error or explain why the unwrap is safe

SarcasticNastik replied:
Refactored code to propagate errors in [b36c116](https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/b36c116babe70817942f2bcd95b97a6d00289ff7)

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -264,6 +349,30 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
     }
 
+    /// Translate `Semantic::Key(key)` to `Semantic::Unsatisfiable` when extracting TapKey
+    pub fn translate_unsatisfiable_pk(&self, key: &Pk) -> Policy<Pk> {
```
This is can consume Self.

SarcasticNastik replied:
Done in [e50766](https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/e507066a5288c09c098d5eac9980066555e287ce).

sanket1729 reviewed (commented):
Did an initial review. I think we would need a few more iterations before we get this.

SarcasticNastik commented on src/policy/semantic.rs:
```diff
@@ -679,10 +679,7 @@ mod tests {
-                1,
-                vec![Policy::KeyHash("".to_owned()), Policy::Older(1000),]
-            )
+            Policy::Threshold(1, vec![Policy::KeyHash("".to_owned()), Policy::Older(1000)])
```
Yes. This code snippet was automatically formatted by `cargo fmt` (version: $1.58.0$).

![image](https://user-images.githubusercontent.com/53688395/149742460-8b5dae61-e9eb-4ca6-9f8e-97d50a6a2568.png)

It was necessary to keep this change to pass the linter.

sanket1729 replied:
I see this is because you removed the "," at the end of all lines. This change is okay, but it is not a part of this commit. But I can live with this

JeremyRubin replied:
i would recommend to bump 1.58.0 and then cargo fmt as one commit maybe?

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -128,6 +137,33 @@ impl fmt::Display for PolicyError {
 impl<Pk: MiniscriptKey> Policy<Pk> {
+    /// Single-Node compilation
+    #[cfg(feature = "compiler")]
+    fn compile_huffman_taptree(policy: &Self) -> Result<TapTree<Pk>, Error> {
```
Right. The naming convention for the function was kept keeping future in mind. I'll rename it to `compile_leaf_taptree` for proper naming convention for this commit.

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -137,30 +143,109 @@ impl fmt::Display for PolicyError {
+            .pop()
+            .expect("huffman tree algorithm is broken")
+            .1;
+        Ok((*node).clone())
```
`node` is of type `Arc<_>` while our return type is `TapTree<Pk>`.

![image](https://user-images.githubusercontent.com/53688395/149748621-2a5334c5-a028-440b-9dc7-7fbdf98edb3f.png)

As I understand, we can't move the value out of `Arc` and the only solution is to call `clone` on the reference for node, either directly in the function or at a later stage (since we require owned value of the node as `Option<TapTree<Pk>>` for `Descriptor::new_tr()` rather than a reference to it.)

![image](https://user-images.githubusercontent.com/53688395/149748931-76347e4b-48b7-4b75-8131-b08dacb9851d.png)

sanket1729 replied:
suggested how to fix it above.

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -128,6 +137,33 @@ impl fmt::Display for PolicyError {
+    }
+    /// Extract the Taproot internal_key from policy tree.
+    #[cfg(feature = "compiler")]
+    fn extract_key(policy: &Self, unspendable_key: Option<Pk>) -> Result<(Pk, &Policy<Pk>), Error> {
```
This would require us to make the `compile_tr` function consume `self` which doesn't align with the `compile<Pk, Ctx>(&self)` function. I will update this function as suggested and use a cloned copy for the same until we discuss further.

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -137,30 +143,109 @@ impl fmt::Display for PolicyError {
+    // TODO: 1. Can try to push the maximum of scaling factors and accordingly update later for
+    // TODO: 1. integer metric. (Accordingly change metrics everywhere)
+    #[cfg(feature = "compiler")]
+    fn flatten_policy(&self, prob: f64) -> Vec<(f64, Policy<Pk>)> {
```
Refactored in [e507066](https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/e507066a5288c09c098d5eac9980066555e287ce)

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -137,30 +143,109 @@ impl fmt::Display for PolicyError {
+        f: T,
+    ) -> Result<TapTree<Pk>, Error>
+    where
+        T: Fn(OrdF64) -> OrdF64,
```
The function `f` refers to the metric we want to optimize the huffman tree over (for example: *leaf probabilities*, *script size* etc. and any combination of those).
For now this function (/signature) is just a dummy (with just *leaf probabilities*) until we definitively decide over the metrics.

sanket1729 replied:
Let's not have this confusing argument for now. And hard-code whatever dummy metric we have

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -32,9 +32,15 @@ use miniscript::types::extra_props::TimeLockInfo;
-use policy::compiler::CompilerError;
+use std::collections::BinaryHeap;
+#[cfg(feature = "compiler")]
+use std::sync::Arc;
```
Refactored in [b36c11](https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/b36c116babe70817942f2bcd95b97a6d00289ff7)

SarcasticNastik commented on src/policy/compiler.rs:
```diff
@@ -38,7 +38,7 @@ type PolicyCache<Pk, Ctx> =
 ///Ordered f64 for comparison
 #[derive(Copy, Clone, PartialEq, PartialOrd, Debug)]
-struct OrdF64(f64);
+pub struct OrdF64(pub f64);
```
Updated in [e507066](https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/e507066a5288c09c098d5eac9980066555e287ce)

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -32,8 +36,13 @@ use policy::compiler;
+#[cfg(feature = "compiler")]
 use Miniscript;
+#[cfg(feature = "compiler")]
+use Tap;
```
Refactored in https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/b36c116babe70817942f2bcd95b97a6d00289ff7

SarcasticNastik commented on src/policy/compiler.rs:
```diff
@@ -994,7 +995,9 @@ where
                 .collect();
 
-            if key_vec.len() == subs.len() && subs.len() <= MAX_PUBKEYS_PER_MULTISIG {
+            if Ctx::sig_type() == SigType::Schnorr {
```
Refactored in https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/b36c116babe70817942f2bcd95b97a6d00289ff7

SarcasticNastik commented on src/policy/mod.rs:
```diff
@@ -233,8 +233,14 @@ mod tests {
+    use std::sync::Arc;
     use DummyKey;
+    #[cfg(feature = "compiler")]
+    use {Descriptor, Tap};
```
Refactored in https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/b36c116babe70817942f2bcd95b97a6d00289ff7

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -128,6 +137,33 @@ impl fmt::Display for PolicyError {
+        let (internal_key, policy) = Self::extract_key(self, unspendable_key).unwrap();
+        let tree = Descriptor::new_tr(
+            internal_key,
+            Some(Self::compile_huffman_taptree(policy).unwrap()),
```
Updated code as suggested in https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/e507066a5288c09c098d5eac9980066555e287ce

sanket1729 commented on src/policy/compiler.rs:
```diff
@@ -994,18 +995,25 @@ where
+                SigType::Schnorr => {
+                    insert_wrap!(AstElemExt::terminal(Terminal::MultiA(k, key_vec)))
+                }
+                _ => {
```
match the `SigType::Ecdsa` to ensure that we don't run any other unnecessary code with SigType is added with another enum variant.

SarcasticNastik replied:
Updated in https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/1a1d19262b859c506d8a60baa3a5ac3643b0fe3f

sanket1729 commented on src/policy/compiler.rs:
```diff
@@ -994,18 +995,25 @@ where
+                    }
+                    // Not a threshold, it's always more optimal to translate it to and()s as we save the
+                    // resulting threshold check (N EQUAL) in any case.
+                    else if k == subs.len() {
```
We could also add this optimization to multi_a. If `k == n`, compiling as `ands() ` is more efficient.

SarcasticNastik replied:
Updated in https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/1a1d19262b859c506d8a60baa3a5ac3643b0fe3f

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -126,27 +132,115 @@ impl fmt::Display for PolicyError {
+            let (p2, s2) = node_weights.pop().expect("len must atleast be two");
+
+            let p = (p1.0).0 + (p2.0).0;
+            node_weights.push((Reverse(OrdF64(p)), Arc::from(TapTree::Tree(s1, s2))));
```
This should not create an `Arc`. The vec should contain all the owned TapTree. You can then do the following:

```diff
- node_weights.push((Reverse(OrdF64(p)), Arc::from(TapTree::Tree(s1, s2))));
+ node_weights.push((Reverse(OrdF64(p)), TapTree::Tree(Arc::new(s1), Arc::new(s2))));

SarcasticNastik replied:
Updated in https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/f7c502629c7a9972c762590b5027ff74513f6c5a

SarcasticNastik replied:
Update in https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/1a1d19262b859c506d8a60baa3a5ac3643b0fe3f

sanket1729 reviewed (commented):
Very close. Last few comments.

SarcasticNastik commented:
The Lint+Fuzz test fails here because of [this](https://github.com/rust-fuzz/honggfuzz-rs/issues/61) issue. Locally running the scripts work just fine.

JeremyRubin commented on src/policy/compiler.rs:
```diff
@@ -994,18 +995,24 @@ where
+                    insert_wrap!(AstElemExt::terminal(Terminal::Multi(k, key_vec)))
                 }
+                _ if k == subs.len() => {
+                    let mut policy = subs.first().expect("No sub policy in thresh() ?").clone();
```
you can use "reduce" here, if the MSRV is 1.51 or higher.

otherwise something like
```
let mut it = policy.iter();
let first = self.next()?;
Some(self.fold(first, |(x,y)| Concrete::And(vec![x,y])))
```

SarcasticNastik replied:
Right! This does seem more idiomatic.

JeremyRubin commented:
shouldn't this try to compile the common: And(Or(a,b), Or(c,d)) to: ac ad bc bd?

how do we judge if the increased number of nodes vs script length cutoff makes sense?

It seems like there is a script length beyond which it always makes sense, maybe like 32 byte script?

SarcasticNastik commented:
@JeremyRubin You're right in the intuitive analysis of the method taproot compiler should ideally adopt. I plan on introducing this type to reduction/ enumeration of policy nodes in a follow-up PR. The reason to keep that on hold was to understand if doing so would turn out to be intractable. For example:- Consider a generic `thresh(k, ...)` k-of-n keys scenario. Enumerating this as separate nodes in disjunction results in nCk (exponential in k). 
In addition, this PR just serves to provide one of the many compilation methods we plan to try/test upon. Another version of this compiler is underway which uses a heuristic to re-arrange the policy nodes and only compile the nodes (`A` and `B`) separately if resulting expected total cost (script_size + average satisfaction cost) of compilation into a taptree is lesser than compilation of a combination of the two (`Or(A, B)`).

dr-orlovsky commented:
@sanket1729 is there plans of getting this into the release?

sanket1729 commented:
@dr-orlovsky, no this won't go into the release. There is still some work to be done here, and the compiler implemented here is feature complete.  We still have a few things we want before we support this. 

Maybe we will have another release with support for partial descriptors, unspendable key support, musig descriptors etc... Possibly 1 month from now

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -126,27 +133,137 @@ impl fmt::Display for PolicyError {
+        Ok(taptree)
+    }
+
+    /// Extract the internal_key from policy tree.
```
The core logic for *internal-key extraction* has been changed from selecting first encountered key (satisfying the policy) to the **most-probable key** (satisfying the policy).

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -126,27 +133,137 @@ impl fmt::Display for PolicyError {
+            let mut prob = 0.;
+            let semantic_policy = self.lift()?;
+            let concrete_keys = self.keys();
+            let key_prob_map: HashMap<_, _> = self
```
The logic of *internal-key extraction* has been changed from selecting **first-encountered key** satisfying the policy to **most-probable key** satisfying the policy.

sanket1729 replied:
I see the ertract key is updated in next commit. No need to fix it here.

sanket1729 commented:
Can you rebase on top of master.? And remove the last CI commit.

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -126,27 +133,137 @@ impl fmt::Display for PolicyError {
-    fn compile_leaf_taptree(&self) -> Result<TapTree<Pk>, Error> {
-        let compilation = self.compile::<Tap>().unwrap();
-        Ok(TapTree::Leaf(Arc::new(compilation)))
+    fn with_huffman_tree<T>(
```
why is this function in `impl Policy`?

SarcasticNastik replied:
Right, I forgot to move the function out of the `impl` block. Will update this.

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -126,27 +133,137 @@ impl fmt::Display for PolicyError {
     }
 
     /// Compile the [`Tr`] descriptor into optimized [`TapTree`] implementation
+    // TODO: We might require other compile errors for Taproot. Will discuss and update.
```
Elaborate on the word optimized here.

SarcasticNastik replied:
Ig the comment here isn't apt here. Will change it altogether!

SarcasticNastik commented:
Changes in the force update:

1. Updated documentation for some functions.
2. Moved `with_huffman_tree` out of the `impl` block.
3. Seperate `internal_key_extraction` and taptree compilation into different commits.

sanket1729 reviewed (commented):
Looks mostly good

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -127,7 +133,30 @@ impl fmt::Display for PolicyError {
 
 impl<Pk: MiniscriptKey> Policy<Pk> {
-    /// Single-Node compilation
+    /// Flatten the [`Policy`] tree structure into a Vector with corresponding leaf probability
```
Done in 3e9214c. Kindly review it and let me know.

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -128,6 +127,30 @@ impl fmt::Display for PolicyError {
+
+    /// Extract the Taproot internal_key from policy tree.
+    #[cfg(feature = "compiler")]
+    fn extract_key(&self, unspendable_key: Option<Pk>) -> Result<(Pk, &Policy<Pk>), Error> {
```
In the first commit 6fb1b29a46ad10231f50f490300ee22a0a66fc07: Is extract key not completely implemented?

SarcasticNastik replied:
Right, this might seem confusing. Will remove it altogether in the first commit.

SarcasticNastik replied:
Resolved in 2b136940

sanket1729 commented:
Sorry, @SarcasticNastik this would require a rebase again :( . 

@apoelstra requesting your review. This is a major contribution from a new contributor and I don't want to go ahead without your review. The commit messages and PR description are helpful in understanding how this is structured.

SarcasticNastik commented:
@sanket1729 I've rebased the PR onto current master. @apoelstra A kind request for your review.

sanket1729 reviewed (approved):
ACK 2b1369409adb49c838253572a59d05b00060df23 . Let's push forward on this. In the interest of ACKs and multiple rebases already done here, I am merging this. We recently shifted to rust 2018. And this code has a bunch of warnings because of it.
